enhancements/installer/assisted-installer-in-cluster.md

```diff
@@ -0,0 +1,641 @@
+### User Stories
+
+#### Day 2 Add Worker
+
+After deploying a cluster with Assisted Installer, I can add workers from bare
+metal hardware on day two using a similar workflow and tool set. I can either
+obtain the discovery live ISO and use my own methods to boot it, or I can use
+the baremetal-operator to boot the live ISO automatically.
+
+#### Multi-cluster
+
+As a user of Red Hat's multi-cluster mangement tools, I can use the assisted
+installation agent-based workflow to create clusters from a pool of bare metal
+inventory.
+
+### Implementation Details/Notes/Constraints
+
+#### Assisted Installer k8s-native API
+
+The Assisted Installer has a [REST
+API](https://generator.swagger.io/?url=https://raw.githubusercontent.com/openshift/assisted-service/master/swagger.yaml)
+that is available at cloud.redhat.com. The service is implemented with a
+traditional relational database. In order to integrate with OpenShift
+infrastructure management tooling, it is necessary to additionally expose its
+capabilities as Kubernetes-native APIs. They will be implemented as Custom
+Resource Definitions and a controller with the assisted service's capabilities.
+
+For the first deliverable, a local deployment of the backing service will
+include new controllers that will operate to achieve the desired state as
+declared in CRDs. Thus the service will expose both REST and Kubernetes-native
+APIs, and both API layers will call the same set of internal methods. The local
+service will include a database; sqlite if possible, or else postgresql.
+
+    --------------------------
+    |  REST API  |  k8s API  |
+    --------------------------
+    | Assisted Service logic |
+    --------------------------
+        |               |
+        v               v
+      SQL DB        File System
+                      storage
+
+For modest clusters that do not have persistent storage available but need
+day-2 ability to add nodes, the database will need to be reconstructable in
+case its current state is lost. Thus the CRs will be the source of truth for
+the specs, and the controller will upon statup first ensure that the DB
+reflects the desired state in CRs. The source of truth for the status, however,
+is the actual state of the agents running on the hosts being installed and not
+necessarily what was previously recorded.
+
+Agents that are running on hardware and waiting for installation-related
+instructions will continue to communicate with the assisted service using the
+existing REST API. In the future that may transition to using a CRD.
+
+**AgentDrivenInstallationImage**
+This resource represents a discovery image that should be used for booting
+hosts. In the REST API this resource is embedded in the "cluster" resource, but
+for kubernetes-native is more natural for it to be separate.
+
+Spec:
+ * clusterName: string
+ * sshPublicKey: string
+ * pullSecretName: string (reference to k8s secret)
+ * ignitionOverrides: object
+ * staticIpConfiguration: object
```

Comment:
I'd like to see some discussion about this design versus likely future directions around supporting static IP configs e.g. #540

For example, consider the standalone cluster case - if the static IP config is stored in a MachineConfig, and you want to create an installation image for a given machine ... wouldn't it make sense for the AgentDrivenInstallationImage to reference that MachineConfig?

Reply:
Note that #540 is mainly about how to configure vlans/bonding - this is potentially possible via MachineConfig because it's likely to be a common configuration for all Machines - there's no way to define per-Machine MachineConfigs ref 
https://github.com/openshift/machine-config-operator/issues/1720

Until/unless that's resolved I don't think MachineConfig provides a solution for static ips (I guess in theory it'd work for single-node deployments, but we want a common interface for multi-node deployments too, right?)

This does raise an interesting question of the proliferation of network interface config interfaces we're exposing users to though, e.g:

1. NetworkManager keyfiles injected via `coreos-install --copy-network` (potentially per-node for UPI/AI but no IPI platforms use this)
2. Ignition customization (potentially per-node in the UPI/AI case)
3. Some future machine-specific MachineConfig interface
4. The declarative kubernetes NMState work being done for day-2 networking (and long term it'd be really nice if day1 and day2 interfaces were the same?)
 
In this case we need some way to configure the install-agent ISO (which IIUC will then pass-through the configuration via coreos-install like in 1 above), which perhaps justifies yet-another interface, but I agree with @markmc it may be worth considering all-the-interfaces that already exist, and justifying if we need a net-new format or some kind of passthrough to one of the existing interfaces.

One simple option might be to allow passing in encoded NM keyfiles via this new API - that provides flexibility (for example bond/vlan support etc without having to add explicit support to a new DSL), but it exposes users to the underlying format (which isn't super-well documented IME).

Reply:
BTW we made the API very similar to what the ignition config expects

Reply:
> BTW we made the API very similar to what the ignition config expects

@avishayt can you clarify that - ignition doesn't have any explicit support for network configuration, only to write files or run systemd units, can you point me to any examples of what the API looks like please?